pkg/api/v1beta3/types.go

```diff
@@ -934,7 +934,7 @@ type ServiceSpec struct {
 	Ports []ServicePort `json:"ports" description:"ports exposed by the service"`
 
 	// This service will route traffic to pods having labels matching this selector. If null, no endpoints will be automatically created. If empty, all pods will be selected.
-	Selector map[string]string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
+	Selector string `json:"selector" description:"label keys and values that must match in order to receive traffic for this service; if empty, all pods are selected, if not specified, endpoints must be manually specified"`
```

Comment:
Sorry it's taken me so long to get to this.

We can't use the string syntax in the API objects. It's not compatible with configuration merges. See: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conventions.md#strategic-merge-patch

There are a number of cases where we need to add labels to both the set of labels and to a selector. Diff and patch aren't feasible if the selector is just one string -- they would require extremely custom logic for this format.

We need a format where the label keys can be used as merge keys. The existing map format works. An array of requirements would also work. My previous proposal was here: https://github.com/GoogleCloudPlatform/kubernetes/issues/341#issuecomment-74364252

We can't make breaking changes to v1beta3. We could consider doing something different in v1, if necessary.

I would also like to avoid polymorphic decoding, and would like to avoid contradicting our API convention of no maps of complex values. 

I do like the simplicity of the map representation for simple selectors, and don't want to lose that. A list of requirements that defaulted the operator to In would be almost as simple. As we did with the string format, we could add direct support for Equals and NotEquals, with singleton string values, to make it even simpler. We could default to Equals or In depending on whether a string value or set were specified. That would look like:
```
"selectorRequirements": [
  { "key": "app", "stringValue": "redis" },
  { "key": "role", "stringValue": "slave" }
]
```
rather than:
```
"selector": {
  "app": "redis",
  "role": "slave"
}
```

However, if we did want to retain the map, we could AND the map with the list of requirements and keep both rather than normalize them into one list of requirements. In the internal representation, we could additionally build the unified label.Selector during conversion, as a non-json public field, to support evaluation, if that's useful. That's what I was thinking in my previous proposal.